docs/madr/decisions/056-meshservice-migration.md

```diff
@@ -0,0 +1,343 @@
+generation.
+
+#### Migration
+
+1. Upgrade to new version
+1. Create MeshMultiZoneServices for cross-zone traffic.
+  - Potentially create MeshService manually to gradually switch over
+1. Enable MeshService generation to have generate MeshServices
+  and have local zone traffic take over
+
+#### Pros
+
+- One of the easiest, most straightforward option as far as stability and UX goes.
+
+#### Cons
+
+- Will be harder to motivate users to switch at all
+- MeshServices need to be deleted if local zone load balancing should be enabled then disabled
+
+### Make cross-zone the default behavior
+
+The cross-zone behavior would be kept for the old, `<kuma.io/service>.mesh` VIP
+and the ClusterIP by generating Envoy clusters that match the status quo
+generation.
+Users would disable this behavior explicitly to switch to local zone behavior.
+
+This means that the clusters generated by Envoy would no longer change on upgrade,
+preserving the traffic.
+
+It also means that policies targeted to real `MeshServices` also target `.mesh` traffic, which is probably fine.
+
+#### Migration
+
+1. Upgrade to new version
+1. MeshServices are now generated but all traffic stays cross-zone
+1. The user creates MeshMultiZoneServices for traffic that should be cross-zone
+1. Disable cross-zone traffic for MeshServices
+1. Disable old VIP generation
+
+#### Pros
+
+- Users get MeshService, HostnameGenerator advantages without breaking the data
+  plane
+
+#### Cons
+
+- Need to keep using ZoneIngress availableServices
+
+### Generate MeshMultiZoneService
+
+Generate `MeshMultiZoneService` and generate hostnames equal to the
+existing `<name>_<namespace>_svc_<port>` format. Users need to disable this
+autogeneration to get the local zone behavior.
+
+Policy matching would need to be adjusted so that old policies also target
+`MeshMultiZoneService`.
+
+#### Migration
+
+This migration is potentially the trickiest.
+
+1. Upgrade to new version
+1. `MeshMultiZoneService` are now generated and they take over generation of
+   `name_namespace_svc_port` clusters. `MeshServices` are also generated
+
+It's not clear what the path or even the goal is from here.
+
+#### Pros
+
+- Makes explicit the fact that requests are going cross-zone
+- Things look exactly the same as how we want cross-zone services to work
+
+#### Cons
+
+- Ihis would require all zones to be upgraded so that MeshServices are synced
+  and can be matched by MeshMultiZoneService
+- Otherwise the most complex option
+
+### Generate new clusters
+
+Instead of `MeshService` taking over cluster generation, generate separate clusters so that:
+
+- `<kuma.io/service>.mesh` -> Old Kuma VIP -> old cluster
+- `HostnameGenerator` DNS name for the `MeshService` -> New Kuma VIP
+  - this includes synced `MeshServices`
+- `name.namespace.svc.cluster.local` -> ClusterIP -> depends
+
+For example, we could have two formats like `name_namespace_svc_port` and `name_namespace_msvc_port`.
```

Comment:
Probably just `name` and `namespace` might be not ideal, as the name for synced resources contains hash suffix and namespace for these resources is `kuma-system`. Should we use the same 4 fields that we have in ResourceIdentifier – name, namespace, zone, mesh?

Reply:
@lobkovilya take another look